crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,222 @@
+        // these will get re-used when building the actual code actions afterwards
+        let edits = diagnostic_edits(
+            document,
+            snapshot.url(),
+            snapshot.encoding(),
+            document.version(),
+            params.context.diagnostics.into_iter(),
+        )
+        .collect::<Result<Vec<_>>>()?;
+
+        let mut response: types::CodeActionResponse = types::CodeActionResponse::default();
+
+        for action in available_actions {
+            match action {
+                SupportedCodeAction::QuickFix => response.extend(quick_fix(edits.iter())),
+                SupportedCodeAction::FixAll => response.extend(fix_all(edits.iter())),
+                SupportedCodeAction::OrganizeImports => {
+                    todo!("Implement the `source.organizeImports` code action")
+                }
+            }
+        }
+
+        Ok(Some(response))
+    }
+}
+
+fn diagnostic_edits<'d>(
+    document: &'d DocumentRef,
+    url: &'d Url,
+    encoding: PositionEncoding,
+    version: DocumentVersion,
+    diagnostics: impl Iterator<Item = types::Diagnostic> + 'd,
+) -> impl Iterator<Item = crate::server::Result<DiagnosticEdit>> + 'd {
+    diagnostics
+        .map(move |diagnostic| {
+            let Some(data) = diagnostic.data.clone() else {
+                return Ok(None);
+            };
+            let diagnostic_fix: crate::lint::DiagnosticFix = serde_json::from_value(data)
+                .map_err(|err| anyhow::anyhow!("failed to deserialize diagnostic data: {err}"))
+                .with_failure_code(lsp_server::ErrorCode::ParseError)?;
+            let edits = diagnostic_fix
+                .fix
+                .edits()
+                .iter()
+                .map(|edit| types::TextEdit {
+                    range: edit
+                        .range()
+                        .to_range(document.contents(), document.index(), encoding),
+                    new_text: edit.content().unwrap_or_default().to_string(),
+                });
+
+            let document_edits = vec![types::TextDocumentEdit {
+                text_document: types::OptionalVersionedTextDocumentIdentifier::new(
+                    url.clone(),
+                    version,
+                ),
+                edits: edits.map(types::OneOf::Left).collect(),
+            }];
+            Ok(Some(DiagnosticEdit {
+                original_diagnostic: diagnostic,
+                diagnostic_fix,
+                document_edits,
+            }))
+        })
+        .filter_map(Result::transpose)
+}
+
+fn quick_fix<'d>(
+    edits: impl Iterator<Item = &'d DiagnosticEdit> + 'd,
+) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
+    edits.map(|edit| {
+        let code = &edit.diagnostic_fix.code;
+        let title = edit
+            .diagnostic_fix
+            .kind
+            .suggestion
+            .clone()
+            .unwrap_or(edit.diagnostic_fix.kind.name.clone());
```

Comment:
Nit: You can avoid cloning here by using `deref`
```suggestion
        let title = edit
            .diagnostic_fix
            .kind
            .suggestion
            .as_deref()
            .unwrap_or(&edit.diagnostic_fix.kind.name);
```